### PR TITLE
Validation for globalRole field InheritedClusterRoles

### PR DESCRIPTION
**Pull request by MbolotSuse**
## Issue: 
Related to rancher/rancher#42214
 
## Problem
The new field introduced in rancher/rancher#42319 (InheritedClusterRoles) introduces a new mechanism for granting permissions. Escalation checks are necessary in the webhook to prevent privilege escalation.
 
## Solution
This changes existing webhook checks for a few resources:
- GlobalRoleBindings:
  - A separate escalation check is given for InheritedClusterRules. You can only give permissions that you already have through this field
  - On create, the GR that the GRB is referring to can't inherit a locked RoleTemplate
- Global Roles:
  -  A separate escalation check is given for InheritedClusterRules. You can only give permissions that you already have through this field
  - New role templates (i.e. a role template added on update or included on create) can't be locked and must have a context of `cluster`
- ClusterRoleTemplateBindings:
  - An update can't change the label indicating GRB ownership 
  - On create, the label indicating GRB ownership must refer to a valid GRB (i.e. in existence and not deleting)
- RoleTemplates:
  - RoleTemplates can't be deleted if they are referred to by a GlobalRole
  
Note: The various escalation checks on GlobalRoles/GlobalRoleBindings don't apply if the request is only changing the object metadata. This was done because Rancher uses norman lifecycle objects, which add annotations and finalizers to these objects as part of the controller functionality.

Since Rancher uses a service account, it fails escalation checks for global roles which use the new `InheritedClusterRoles` field. Since we don't currently have a clean path to migrate from norman controller to wrangler controllers (which could in theory use the webhook bypass logic), we need to allow changes to the metadata to allow the controllers to work properly.

 
## Testing
Unit tests were added for net-new code. Manual testing is still pending.

## Notes
The first commit refers to my fork of Rancher, which will be necessary until the Rancher-side PR (rancher/rancher#42319) is merged

### CODE REVIEW

File: pkg/resources/management.cattle.io/v3/clusterroletemplatebinding/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/clusterroletemplatebinding/validator.go:182]: I think that validation would be a good idea, but I'm not a fan of implementing it in this PR. I'd say let's address it in a separate issue.

File: pkg/auth/globalrole.go
Comment (MbolotSuse) [reply pkg/auth/globalrole.go:68]: Personally speaking, I don't like how that structure exposes internal resources for usage by other packages. I'd prefer that it offered an interface for performing specific operations that someone would want the cache for. That being said, I'll adjust it for the sake of consistency since this isn't that big of a deal.

File: pkg/resolvers/grbClusterResolver_test.go
Comment (KevinJoiner) [reply pkg/resolvers/grbClusterResolver_test.go:110]: Agreed, it won't increase our test coverage, It just seems helpful to have test for all happy paths incase there are future changes to the implementation

**Review by KevinJoiner — APPROVED** [review]

File: go.mod
Comment (KevinJoiner) [inline go.mod:8]: This PR LGTM, whenever this fork is replacement, can be removed.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go
Comment (pjbgf) [inline pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go:155]: As multiple role templates could be locked, what if the `for` aggregates all the locked RT names into a variable, and after the `for` we return the aggregated list of locked RT names? The reasoning being that the short-circuit does not save much time/space as the check is inexpensive and all the data is already in memory.

File: pkg/resolvers/grbClusterResolver.go
Comment (pjbgf) [inline pkg/resolvers/grbClusterResolver.go:26]:
```suggestion
// which apply to cluster(s). This function can only be called once for each unique instance of grbCache.
```

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (pjbgf) [inline pkg/resources/management.cattle.io/v3/globalrole/validator.go:92]:
```suggestion
	// check for escalation separately between cluster permissions and global permissions to prevent crossover
```

File: pkg/resources/management.cattle.io/v3/roletemplate/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/roletemplate/validator.go:213]: This is something that was previously added. I can change it, but if you could restrict your comments to the changes I would appreciate it, there's a lot here to look at as-is.

File: pkg/resources/management.cattle.io/v3/roletemplate/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/roletemplate/validator.go:231]: This is something that was previously added. I can change it, but if you could restrict your comments to the changes I would appreciate it, there's a lot here to look at as-is.

File: pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go:165]: Yep, that makes sense and I think produces a far better UX.

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/globalrole/validator.go:168]: I generally prefer the else if there since the two if statements are more connected in my mind than not connected (they both evaluate the validity of the target rt). But I can remove it - I would appreciate you pointing me to the style guide that indicates that this is preferred though.

File: pkg/resources/management.cattle.io/v3/roletemplate/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/roletemplate/validator.go]: Done.

File: pkg/resources/management.cattle.io/v3/roletemplate/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/roletemplate/validator.go]: Done.

File: pkg/resources/management.cattle.io/v3/roletemplate/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/roletemplate/validator.go]: Done.

File: pkg/resources/management.cattle.io/v3/roletemplate/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/roletemplate/validator.go]: Done.

File: pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go]: Done.

File: pkg/resolvers/grbClusterResolver.go
Comment (MbolotSuse) [reply pkg/resolvers/grbClusterResolver.go]: Done.

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/globalrole/validator.go]: Done.

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (MbolotSuse) [reply pkg/resources/management.cattle.io/v3/globalrole/validator.go]: Done.

**Review by pjbgf — APPROVED** [review]

**Review by KevinJoiner — APPROVED** [review]

File: go.mod
Comment (macedogm) [reply go.mod:8]: Is this replacement being solved in https://github.com/rancher/rancher/pull/42319?

**Review by macedogm — APPROVED** [review]
LGTM, added only one comment about the replacement in go.mod.

File: go.mod
Comment (MbolotSuse) [reply go.mod:8]: Yep, once that merges I can make this point to a real rancher commit (Rather than my fork like it's doing now). But I need that one merged first.

Comment (macedogm) [reply go.mod:8]: Thanks for the explanation. Anyway, LGTM. 👍🏻

Comment (KevinJoiner) [inline go.mod:8]: Placeholder comment to block merge until rancher PR is merged.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: go.mod
Comment (MbolotSuse) [reply go.mod]: I've made the replacement, this now refers to a real rancher version.

**Review by KevinJoiner — APPROVED** [review]

**Review by macedogm — APPROVED** [review]

**Pull request merged**
